synapse/http/servlet.py

```diff
@@ -572,6 +572,25 @@ def parse_string_from_args(
     return strings[0]
 
 
+@overload
+def parse_json_value_from_request(request: Request) -> JsonDict:
+    ...
+
+
+@overload
+def parse_json_value_from_request(
+    request: Request, *, allow_empty_body: Literal[True]
+) -> Optional[JsonDict]:
+    ...
+
+
+@overload
+def parse_json_value_from_request(
+    request: Request, *, allow_empty_body: bool = False
+) -> Optional[JsonDict]:
```

Comment:
Shouldn't this be 

```suggestion
) -> JsonDict:
```

?

Reply:
The thing is that none of these overrides say that when `allow_empty_body=False` then the return value is `JsonDict`. I think we actually want to change the variant above with `Literal[True]` to be `Literal[False]` and return a `JsonDict`. (You could specify both a `Literal[False]` and `Literal[True]` variant, but the `Literal[True]` variant is the same as the non literal variant and so its redundant).

Reply:
Ah, maybe that's where I messed up then. The suggested change causes mypy to complain about overlapping definitions.

Reply:
I get very confused at the correct way of doing these sort of overrides every single time tbf